portalnet/src/accept_queue.rs

```diff
@@ -0,0 +1,206 @@
+use delay_map::HashMapDelay;
+use ethportal_api::{types::enr::Enr, OverlayContentKey};
+use futures::prelude::*;
+use rand::{seq::SliceRandom, thread_rng};
+use tokio::time::Duration;
+use tracing::{debug, warn};
+
+/// A record of peers that have offered a content key.
+struct SeenPeers {
+    /// The peer that originally offered the content key.
+    origin: Enr,
+    /// A list of subsequent peers that have also offered the content key.
+    fallback: Vec<Enr>,
+}
+
+/// In-memory queue of content keys that have been accepted, and are
+/// currently being processed.
+/// This queue is designed for content keys that have already
+/// been checked with storage that they are within radius.
+// It's possible that some content keys evade tracking through the entire processing
+// cycle, due to error handling inside the OverlayService, or threads
+// panicking / deadlocking. The use of a delay
+// map will remove content keys which have been in the queue for too long or
+// might be stuck, and avoid infinitely-growing queues / indefinitely preventing
+// previously seen content keys from being accepted.
+pub struct AcceptQueue<TContentKey>
+where
+    TContentKey: OverlayContentKey,
+{
+    // a map of content keys actively being transferred / processed
+    // pointing to seen peers that have offered them the content
+    content_key_map: HashMapDelay<TContentKey, SeenPeers>,
+}
+
+const OFFER_QUEUE_TIMEOUT: Duration = Duration::from_secs(120);
+
+impl<TContentKey> Default for AcceptQueue<TContentKey>
+where
+    TContentKey: OverlayContentKey,
+{
+    fn default() -> Self {
+        Self {
+            content_key_map: HashMapDelay::new(OFFER_QUEUE_TIMEOUT),
+        }
+    }
+}
+
+impl<TContentKey> AcceptQueue<TContentKey>
+where
+    TContentKey: OverlayContentKey,
+{
+    /// Tries to add a content key to the queue.
+    /// If the key is not in the queue, it is added to the queue and returns true.
+    /// If the key is in the queue, the seen peer is stored as a fallback, and it returns false.
+    /// Also polls for expired items, which will remove them from the queue.
+    pub fn add_key_to_queue(&mut self, content_key: &TContentKey, peer: &Enr) -> bool {
+        // poll for expired items, which will remove them from the queue
+        let _ = future::poll_fn(|cx| self.content_key_map.poll_expired(cx)).now_or_never();
+        if let Some(mut seen_peers) = self.content_key_map.remove(content_key) {
+            if seen_peers.origin == *peer || seen_peers.fallback.contains(peer) {
+                warn!(
+                    "Received multiple offers containing the same content key: {content_key} from peer: {peer}"
+                );
```

Comment:
If the key is in the queue, we will get inside this block and then jump to line 73 and return `true`. I think in this case we want to return `false`, because the key is already included in the queue?

Reply:
I don't think so, we jump to line `70` not `73` - and from line `70` we re-insert the "removed" `seen_peers` and then return a `false`. This case is also covered by the `test_queue_keeps_record_after_duplicate_offers_from_same_peer` test